src/virtual_dom/patch/patch_gen.rs

```diff
@@ -0,0 +1,474 @@
+    AppendEl {
+        el_new: &'a mut El<Ms>,
+    },
+    AppendText {
+        text_new: &'a mut Text,
+    },
+    InsertEl {
+        el_new: &'a mut El<Ms>,
+        next_node: web_sys::Node,
+    },
+    InsertText {
+        text_new: &'a mut Text,
+        next_node: web_sys::Node,
+    },
+    PatchEl {
+        el_old: El<Ms>,
+        el_new: &'a mut El<Ms>,
+    },
+    PatchText {
+        text_old: Text,
+        text_new: &'a mut Text,
+    },
+    ReplaceElByEl {
+        el_old: El<Ms>,
+        el_new: &'a mut El<Ms>,
+    },
+    ReplaceElByText {
+        el_old: El<Ms>,
+        text_new: &'a mut Text,
+    },
+    ReplaceTextByEl {
+        text_old: Text,
+        el_new: &'a mut El<Ms>,
+    },
+    RemoveEl {
+        el_old: El<Ms>,
+    },
+    RemoveText {
+        text_old: Text,
+    },
+}
+
+#[derive(Clone, Debug, PartialEq, Eq, PartialOrd, Ord)]
+enum PatchKey {
+    Element((Option<Namespace>, Tag, Option<ElKey>)),
+    Text,
+}
+
+fn make_key<Ms: 'static>(node: &Node<Ms>) -> Option<PatchKey> {
+    match node {
+        Node::Element(el) => Some(PatchKey::Element((
+            el.namespace.clone(),
+            el.tag.clone(),
+            el.key.clone(),
+        ))),
+        Node::Text(_) => Some(PatchKey::Text),
+        Node::Empty => None,
+    }
+}
+
+pub(crate) struct PatchGen<'a, Ms, OI, NI>
+where
+    Ms: 'static,
+    OI: Iterator<Item = Node<Ms>>,
+    NI: Iterator<Item = &'a mut Node<Ms>>,
+{
+    old_children_iter: Peekable<OI>,
+    new_children_iter: Peekable<NI>,
+    old_children: VecDeque<Node<Ms>>,
+    new_children: VecDeque<&'a mut Node<Ms>>,
+    matching_child_old: Option<OI::Item>,
+    matching_child_new: Option<NI::Item>,
+    matching_key: Option<PatchKey>,
+    keyed_mode: bool,
+}
+
+impl<'a, Ms, OI, NI> PatchGen<'a, Ms, OI, NI>
+where
+    Ms: 'static,
+    OI: Iterator<Item = Node<Ms>>,
+    NI: Iterator<Item = &'a mut Node<Ms>>,
+{
+    pub fn new(old_children_iter: OI, new_children_iter: NI) -> Self {
+        Self {
+            old_children_iter: old_children_iter.peekable(),
+            new_children_iter: new_children_iter.peekable(),
+            old_children: VecDeque::new(),
+            new_children: VecDeque::new(),
+            matching_child_old: None,
+            matching_child_new: None,
+            matching_key: None,
+            keyed_mode: false,
+        }
+    }
+
+    fn next_command(&mut self) -> Option<PatchCommand<'a, Ms>> {
+        if self.keyed_mode {
+            if self.matching_key.is_none()
+                && (self.old_children_iter.peek().is_some()
+                    || self.new_children_iter.peek().is_some())
+            {
+                self.scan();
+            }
+            if self.matching_key.is_some() {
+                self.yield_keyed()
+            } else {
+                self.yield_remaining()
+            }
+        } else {
+            self.yield_keyless()
+        }
+    }
+
+    fn scan(&mut self) {
```

Comment:
Actually there is a bug.
Pure immutable code in this case will look like State -> State, so mutation in this case reasonable for me.
Anyway I'll fix it.

Reply:
Yeah, I agree that mutation is often a better choice (it's also one of the reasons why I moved to Rust). 
But in combination with `Self` in this case it's pretty confusing. 
I think, for instance, passing only variables that you want to mutate instead of the whole `Self` would be much better - it's clear what the function modifies. And with some doc comments for the function and it's parameters we are good to go.

Reply:
Fixed in new commit.

Reply:
(Let me know once you are ready for the next code review round 🙂)